play-v29/src/main/scala/com/gu/googleauth/groups.scala

```diff
@@ -23,15 +25,34 @@ import scala.jdk.CollectionConverters._
  *
  * @param impersonatedUser a separate domain-user account email address (eg 'example@guardian.co.uk'), the email address
  *                         of the user the application will be impersonating when making calls.
+ * @param serviceAccountCredentials Google OAuth2 credentials.
+ * @param cacheDuration how long to cache each user's groups for (defaults to 1 minute).
+ *
  */
-class GoogleGroupChecker(impersonatedUser: String, serviceAccountCredentials: ServiceAccountCredentials) {
+class GoogleGroupChecker(
+  impersonatedUser: String,
+  serviceAccountCredentials: ServiceAccountCredentials,
+  cacheDuration: Duration = Duration.ofMinutes(1)
+) {
 
   private val googleCredentials: GoogleCredentials = serviceAccountCredentials.createDelegated(impersonatedUser)
 
   private val directoryService: Directory = DirectoryService(googleCredentials, ADMIN_DIRECTORY_GROUP_READONLY)
 
-  def retrieveGroupsFor(userEmail: String)(implicit ec: ExecutionContext): Future[Set[String]] = for {
-    resp <- Future { blocking { directoryService.groups.list.setUserKey(userEmail).execute() } }
-  } yield resp.getGroups.asScala.map(_.getEmail).toSet
-  
+  type Email = String
+  private val cache: LoadingCache[Email, Set[String]] = CacheBuilder.newBuilder()
```

Comment:
Only authed users will end up in the cache, so the unbounded cache size is fine.